racket/src/cs/rumble/intmap.ss

```diff
@@ -63,35 +61,42 @@
         [(Co? t) (length (Co-pairs t))]
         [else 0]))
 
-(define (intmap-ref t key def)
+(define (do-intmap-ref t key with-leaf with-pair default)
   (let ([et (intmap-eqtype t)]
         [root (intmap-root t)])
     (if root
-        ($intmap-ref et root (hash-code et key) key def)
-        ($fail def))))
+        (do-$intmap-ref et root (hash-code et key) key with-leaf with-pair default)
+        default)))
 
-(define ($intmap-ref et t h key def)
-  (cond
-   [(Br? t)
-    (if (fx<= h (Br-prefix t))
-        ($intmap-ref et (Br-left t) h key def)
-        ($intmap-ref et (Br-right t) h key def))]
+(define (do-$intmap-ref et t h key with-leaf with-pair default)
+  (let loop ([t t])
+    (cond
+     [(Br? t)
+      (if (fx<= h (Br-prefix t))
+          (loop (Br-left t))
+          (loop (Br-right t)))]
 
-   [(Lf? t)
-    (if (key=? et key (Lf-key t))
-        (Lf-value t)
-        ($fail def))]
+     [(Lf? t)
+      (if (key=? et key (Lf-key t))
+          (with-leaf t)
```

Comment:
Does replacing the known selector used with a call to an unknown function affect performance? I would not expect `do-$intmap-ref` to be inlined, but I may guess wrong.

Reply:
Performance seems minimally effected by this change. 

To give you some more context, originally I implemented `do-$intmap-ref` as syntax, after the model of [`lookup-keyval` in Chez's implementation of mutable eq-hashtables](https://github.com/cisco/ChezScheme/blob/58663d92a353969e3b27de2e1a44df571036e1b2/s/library.ss#L1380). However, this had a marked negative effect on the "large table" tests from `hash-demo`:

Original implementation:
```
large tables
cpu time: 123 real time: 123 gc time: 2   MB: 292
cpu time: 162 real time: 162 gc time: 2   MB: 292
```

With `do-$intmap-ref` as syntax:
```
large tables
cpu time: 162 real time: 162 gc time: 2   MB: 292
cpu time: 336 real time: 336 gc time: 2   MB: 292
```

With `do-$intmap-ref` as function:
```
large tables
cpu time: 125 real time: 125 gc time: 2   MB: 292
cpu time: 164 real time: 164 gc time: 3   MB: 292
```

The other tests in that demo, however, are unaffected. I have no idea what to make of this. If Chez does loop unrolling, then I could see how the increased code size might have some negative cache effects, but I don't think it does any unrolling.

Reply:
Well, I guess the [unrolling story](https://github.com/cisco/ChezScheme/blob/58663d92a353969e3b27de2e1a44df571036e1b2/s/cp0.ss#L79) is a bit more complicated than that.